x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/InitialNodeSecurityAutoConfiguration.java

```diff
@@ -33,6 +36,7 @@
 public class InitialNodeSecurityAutoConfiguration {
 
     private static final Logger LOGGER = LogManager.getLogger(InitialNodeSecurityAutoConfiguration.class);
+    private static final BackoffPolicy BACKOFF_POLICY = BackoffPolicy.exponentialBackoff();
```

Comment:
Given the lack of a better/more appropriate hook to trigger the token generation and the (admittedly only presumed in this point) scarcity of the manifestation of this, I think that the simplicity of a retry-with-backoffs solution is sufficient. Happy to discuss alternatives though

Reply:
I took a close look at this.

I think the race is because the enrollment token generation starts when the `Node` is **instantiated** but the"tcp" and "http" transports, which are needed for the token to actually be created, only become available when the node **started**.

> Given the lack of a better/more appropriate hook to trigger the token generation

Ideally, the token generation should start when the state has been recovered without a `security` index (what we do today) AND the `ClusterPlugin#onNodeStarted` has been invoked on the `Security` plugin (at which point we know the transports are available). Basically, we are safe to trigger the generation of the enrollment token, when the last of these two actions completed (I thought of using a `CountDown(2)` for it).

I tried having a quick attempt at it to suggest here, but I couldn't get it to a small code change size, because the "when both of the two action completed" condition requires shifting around every line in the class `InitialNodeSecurityAutoConfiguration`.

If you agree with my train of thoughts above, I would suggest we open an enhancement request to avoid the polling you're introducing here (because we're in a bit of a hurry atm).

Reply:
Yes, I do agree both with your suggestion on the proposed solution and to merge this as a stop-gap solution until we can figure out the implementation details. I was looking on `onNodeStarted` too when thinking about solutions, but it wasn't obvious to me either how to best make this change so that our action is triggered when both these are notified. Thanks for spending the time on proper possible solutions

Reply:
I opened https://github.com/elastic/elasticsearch/issues/81160, happy to review if you get the cycles/time to look at it, otherwise I'll try to plan this for on of the following releases